mcs/class/Mono.Posix/Mono.Unix.Native/Syscall.cs

```diff
@@ -4356,6 +4907,542 @@ public static long pwritev (int fd, Iovec[] iov, long offset)
+		static extern int sys_socket (int domain, int type, int protocol);
+
+		public static int socket (AddrFamily domain, SockType type, SockFlags flags, SockProtocol protocol)
+		{
+			var _domain = NativeConvert.FromAddrFamily (domain);
+			var _type = NativeConvert.FromSockType (type);
+			var _flags = NativeConvert.FromSockFlags (flags);
+			// protocol == 0 is a special case (uses default protocol)
+			var _protocol = protocol == 0 ? 0 : NativeConvert.FromSockProtocol (protocol);
+
+			return sys_socket (_domain, _type | _flags, _protocol);
+		}
+
+		public static int socket (AddrFamily domain, SockType type, SockProtocol protocol)
+		{
+			return socket (domain, type, 0, protocol);
+		}
+
+		// socketpair(2)
+		//    int socketpair(int domain, int type, int protocol, int sv[2]);
+		[DllImport (MPH, SetLastError=true, 
+				EntryPoint="Mono_Posix_Syscall_socketpair")]
+		static extern int sys_socketpair (int domain, int type, int protocol, out int sv0, out int sv1);
+
+		public static int socketpair (AddrFamily domain, SockType type, SockFlags flags, SockProtocol protocol, out int sv0, out int sv1)
+		{
+			var _domain = NativeConvert.FromAddrFamily (domain);
+			var _type = NativeConvert.FromSockType (type);
+			var _flags = NativeConvert.FromSockFlags (flags);
+			// protocol == 0 is a special case (uses default protocol)
+			var _protocol = protocol == 0 ? 0 : NativeConvert.FromSockProtocol (protocol);
+
+			return sys_socketpair (_domain, _type | _flags, _protocol, out sv0, out sv1);
+		}
+		
+		public static int socketpair (AddrFamily domain, SockType type, SockProtocol protocol, out int sv0, out int sv1)
+		{
+			return socketpair (domain, type, 0, protocol, out sv0, out sv1);
+		}
+
+		public static int socketpair (AddrFamily domain, SockType type, SockFlags flags, SockProtocol protocol, int[] socket_vector)
+		{
+			if (socket_vector == null || socket_vector.Length != 2) {
+				SetLastError (Errno.EINVAL);
+				return -1;
+			}
+			int sv0, sv1;
+			int r = socketpair (domain, type, flags, protocol, out sv0, out sv1);
+			socket_vector[0] = sv0;
+			socket_vector[1] = sv1;
+			return r;
+		}
+
+		public static int socketpair (AddrFamily domain, SockType type, SockProtocol protocol, int[] socket_vector)
+		{
+			return socketpair (domain, type, 0, protocol, socket_vector);
+		}
+		
+		// sockatmark(2)
+		//    int sockatmark(int sockfd);
+		[DllImport (LIBC, SetLastError=true)]
+		public static extern int sockatmark (int socket);
+		
+		// listen(2)
+		//    int listen(int sockfd, int backlog);
+		[DllImport (LIBC, SetLastError=true)]
+		public static extern int listen (int socket, int backlog);
+		
+		// bind(2)
+		//    int bind(int sockfd, const struct sockaddr *addr, socklen_t addrlen);
+		[DllImport (MPH, SetLastError=true, 
+				EntryPoint="Mono_Posix_Syscall_bind")]
+		public static unsafe extern int bind (int socket, byte* addr, long addrlen);
+
+		public static unsafe int bind (int socket, byte[] addr, long addrlen)
```

Comment:
Please make this and the above `byte*` overload `private`. We shouldn't expect anybody to use these overloads.